### PR TITLE
Query history from history server

### PR DESCRIPTION
**Pull request by progval**
test suite here: https://github.com/progval/irctest/pull/292

### CODE REVIEW

File: sable_history/src/pg_history_service.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                    Ok(HistoryLogEntry {
+                        id: (),
+                        details: NetworkStateChange::NewMessage(update::NewMessage {
+                            message: (),
+                            source: (),
+                            target: (),
+                        }),
+                        source_event: (),
+                        timestamp: (),
+                    })
```
Comment (progval) [inline sable_history/src/pg_history_service.rs:152]: @spb Could you fill the blanks here?

Comment (spb) [reply sable_history/src/pg_history_service.rs:152]: I don't think they can be filled in as-is - the IDs might not be meaningful to the requesting server if those messages have aged out of local retention. I think the interface needs modifying to return the information directly instead of by ID reference.

Comment (progval) [reply sable_history/src/pg_history_service.rs:152]: I can work on those later after I know exactly which ones need to be removed

Comment (spb) [reply sable_history/src/pg_history_service.rs:152]: OK, so in order of the above code:

- HistoryLogEntry ID isn't needed
- Message content, source, and target will be required, but will need to have the actual content and not IDs
- Source event ID probably isn't useful. We might need some information from the event to fill in capability-based tags, but the event itself will be long gone so those fields will need to be stored in the history db as their own separate fields
- Timestamp will be needed for server_time at least

...but honestly I think starting from here and trying to remove things is the wrong way to think about it. We know what information the chathistory handler needs to be able to send the messages to the client, so the better answer is to just write a struct that contains that information in the most logical manner, then make sure that the database contains that information and that the local history service can translate from the in-memory log into that format.

Comment (progval) [reply sable_history/src/pg_history_service.rs:152]: "We know what information the chathistory handler needs to be able to send the messages to the client" -> currently, the chathistory handler reuses existing code through `ClientServer::send_item`; adding a new struct means we need to either duplicate or rewrite everything in `sable_ircd/src/messages/send_history.rs`

Comment (progval) [reply sable_history/src/pg_history_service.rs:152]: Ah, or we could directly build the same `OutboundClientMessage` that `ClientServer::send_item` would before sending, hmm

Comment (spb) [reply sable_history/src/pg_history_service.rs:152]: Right. The new chathistory command handler won't be able to rely on `send_item` because it can't use a `HistoryLogEntry`, because the historic user state that it would need to refer to doesn't exist any more. To support remote history it'll need to construct the outbound client message directly, including getting the tag data from the history message provided by the history server and using string data instead of user/channel objects for the source and target.

File: sable_network/src/history/local_service.rs [hunk 2]
```diff
@@ -103,20 +105,47 @@ impl<'a> LocalHistoryService<'a> {
...
+                            source,
+                            target: _, // assume it's the same as the argument we got
+                        }) => Some(HistoricalEvent::Message {
+                            id: *message,
+                            message_type: crate::network::state::MessageType::Notice, // TODO
+                            source: "".to_string(),                                   // TODO
+                            source_account: None,                                     // TODO
+                            target,
+                            text: "".to_string(), // TODO
+                        }),
```
Comment (progval) [inline sable_network/src/history/local_service.rs:135]: @spb More blanks to fill

Comment (spb) [reply sable_network/src/history/local_service.rs:135]: At this point you need to look up the actual message by ID in the network state and use the contents to fill in these fields.

Comment (progval) [reply sable_network/src/history/local_service.rs:135]: I know but I was hoping you would do it for me :)

I'll get to it after I'm done with the rest if you're busy

Comment (spb) [reply sable_network/src/history/local_service.rs:135]: See 9731d9b0baab47bf9920adb960803a68e875e2dc for most of it

File: sable_network/src/history/service.rs [hunk 3]
```diff
@@ -98,5 +98,17 @@ pub trait HistoryService {
...
         target: TargetId,
         request: HistoryRequest,
-    ) -> impl Future<Output = Result<impl IntoIterator<Item = HistoryLogEntry>, HistoryError>> + Send;
+    ) -> impl Future<Output = Result<impl IntoIterator<Item = HistoricalEvent>, HistoryError>> + Send;
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub enum HistoricalEvent {
+    Message {
+        id: Uuid7,
```
Comment (spb) [inline sable_network/src/history/service.rs:107]: Should be MessageId

File: sable_network/src/history/service.rs [hunk 4]
```diff
@@ -98,5 +98,17 @@ pub trait HistoryService {
...
+        source: String,
+        source_account: Option<String>,
+        target: TargetId,
```
Comment (spb) [inline sable_network/src/history/service.rs:110]: Needs to be a String - the thing referenced by the Id might not exist any more

File: sable_ircd/src/command/handlers/chathistory.rs [hunk 5]
```diff
@@ -197,20 +198,48 @@ async fn list_targets<'a>(
...
+                text,
+            } => {
+                let (timestamp, _) = id
+                    .get_timestamp()
+                    .expect("message has non-v7 UUID")
+                    .to_unix();
+                let msg = message::Message::new(&source, &target, message_type, &text)
+                    .with_tag(server_time::server_time_tag(
+                        i64::try_from(timestamp).unwrap_or(i64::MAX),
+                    ))
```
Comment (progval) [inline sable_ircd/src/command/handlers/chathistory.rs:228]: This is unsound. I stumbled upon this issue after running irctest a dozen times:

1. client sends a message, gets this echo: `@time=2025-10-27T17:18:50.000Z;msgid=019a26ae-048f-7352-80cb-82d9d3ded30f :bar!username@localhost PRIVMSG #chanac0a9c4d540ce04f7a35da65 :this is message 3` (note that `019a26ae-048f-7352-80cb-82d9d3ded30f`'s timestamp is `2025-10-27T17:18:49.999Z`)
2. sable_history writes this: `Message { id: 019a26ae-048f-7352-80cb-82d9d3ded30f, source_user: 1, target_channel: 770697994130427908, text: "this is message 3", message_type: Privmsg }`
3. client asks for chathistory
4. ircd asks sable_history, gets `Message { id: MessageId(Uuid7(019a26ae-048f-7352-80cb-82d9d3ded30f)), source: "bar!username@localhost", source_account: None, target: "#chanac0a9c4d540ce04f7a35da65", message_type: Privmsg, text: "this is message 3" }` back
5. ircd replies this to the client: `@time=2025-10-27T17:18:49.000Z;msgid=019a26ae-048f-7352-80cb-82d9d3ded30f;batch=e5286d57605d2baa2395dc78c4d7d042 :bar!username@localhost PRIVMSG #chanac0a9c4d540ce04f7a35da65 :this is message 3`

Comment (progval) [reply sable_ircd/src/command/handlers/chathistory.rs:228]: > IMO we should build the UUID in a way that ensures it matches the timestamp.

except we can't because the UUID is part of the message, which is built before we have a timestamp from the NewMessage state change

Comment (progval) [reply sable_ircd/src/command/handlers/chathistory.rs:225]: Fixed by storing the update timestamp as its own column in the DB

**Comment by progval** [comment]
@spb AFAICT, I only need to implement AROUND implementation before this PR is ready to be merged. Could you start reviewing?

File: sable_history/src/pg_history_service.rs [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    ) -> Result<impl IntoIterator<Item = HistoricalEvent>, HistoryError> {
+        // TODO: access control
+        let TargetId::Channel(channel_id) = target else {
+            // TODO: PMs
+            return Err(HistoryError::InvalidTarget(target));
+        };
+
+        let mut connection_lock = self.database_connection.lock().await;
+
+        let db_channel_id = i64::try_from(channel_id.as_u64()).expect("channel id overflows u64");
```
Comment (spb) [inline sable_history/src/pg_history_service.rs:90]: I've been preferring `as` conversions for this (to be replaced eventually with cast_signed/cast_unsigned), to just store and retrieve the bit pattern on the basis that we don't need the database server to be ordering IDs, just checking equality. 

Unless we do need the database to order them, in which case we need to change the existing code (and should probably enforce on creation that an id fits in 63 bits). Either way I think it's better to be consistent.

File: sable_history/src/pg_history_service.rs [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+            ))
+            .load_stream(&mut *self.database_connection.lock().await)
+            .await
+        {
+            Err(e) => {
+                tracing::error!("Could not get history channels: {e}");
+                return HashMap::new();
+            }
+            Ok(rows) => {
+                rows.map(|row| row.expect("Could not deserialize row"))
```
Comment (spb) [inline sable_history/src/pg_history_service.rs:53]: Is panicking appropriate here or should we log an error and skip instead? We've got a lot of existing PoC code that panics all over the place but maybe it's time to make more efforts not to.

File: sable_history/src/pg_history_service.rs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+            .select((
+                messages::dsl::id,
+                messages::dsl::timestamp,
+                messages::dsl::message_type,
+                messages::dsl::text,
+                historic_users::dsl::nick,
+                historic_users::dsl::ident,
+                historic_users::dsl::vhost,
+                historic_users::dsl::account_name,
+            ))
```
Comment (spb) [inline sable_history/src/pg_history_service.rs:113]: Is it worth packaging this query up into a helper function in the models?

File: sable_history/src/server/mod.rs [hunk 9]
```diff
@@ -27,7 +29,7 @@ pub struct HistoryServerConfig {
 pub struct HistoryServer {
     node: Arc<NetworkNode>,
     history_receiver: Mutex<UnboundedReceiver<sable_network::rpc::NetworkHistoryUpdate>>,
-    database_connection: Mutex<AsyncPgConnection>,
+    database_connection: Mutex<AsyncPgConnection>, // TODO: use a connection pool
```
Comment (spb) [inline sable_history/src/server/mod.rs:32]: Yes, definitely

File: sable_network/Cargo.toml [hunk 10]
```diff
@@ -21,6 +21,7 @@ pretty_assertions = "1.4"
 [dependencies]
 sable_macros = { path = "../sable_macros" }
 
+async-trait = "0.1.83"
```
Comment (spb) [inline sable_network/Cargo.toml:24]: Why?

File: sable_network/src/history/local_service.rs [hunk 11]
```diff
@@ -20,12 +22,12 @@ fn target_id_for_entry(for_user: UserId, entry: &HistoryLogEntry) -> Option<Targ
 }
 
 /// Implementation of [`HistoryService`] backed by [`NetworkNode`]
-pub struct LocalHistoryService<'a> {
-    node: &'a NetworkNode,
+pub struct LocalHistoryService<'a, NetworkPolicy: policy::PolicyService> {
```
Comment (spb) [inline sable_network/src/history/local_service.rs:25]: I'm starting to really question whether making Node generic over its policy was a good idea. Not much to do about it in this PR but it wants thought more generally.

File: sable_network/src/history/mod.rs [hunk 12]
```diff
@@ -22,3 +26,7 @@ impl HistoryItem for HistoryLogEntry {
         &self.details
     }
 }
+
+/// A more concrete representation of [`HistoryItem`], with all its fields inflated
+/// to strings that will be sent to the client
+pub enum HistoryMessage {}
```
Comment (spb) [inline sable_network/src/history/mod.rs:32]: Doesn't seem to be used anywhere?

File: sable_history/src/pg_history_service.rs (hunk 7 above)
Comment (progval) [reply sable_history/src/pg_history_service.rs:53]: Based on https://docs.rs/diesel/latest/diesel/result/enum.Error.html this happens when:

* there is a major logic error on our side,
* the database misbehaved, or
* the database schema does not match Rust's

It seems reasonable to crash either way.

File: sable_history/src/pg_history_service.rs (hunk 6 above)
Comment (progval) [reply sable_history/src/pg_history_service.rs:90]: ok for now, but in the future we should do it explicitly, eg. with bytemuck. Behavior of `as` depends on compilation profile, so it shouldn't be relied upon.

File: sable_network/src/history/mod.rs (hunk 12 above)
Comment (progval) [reply sable_network/src/history/mod.rs]: indeed, I replaced it with `HistoricalEvent`

File: sable_history/src/pg_history_service.rs (hunk 7 above)
Comment (progval) [reply sable_history/src/pg_history_service.rs]: Heh, you're right. I removed the panics.

File: sable_history/src/pg_history_service.rs [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+            .filter(messages::dsl::target_channel.eq(db_channel_id));
+        Ok(match request {
+            HistoryRequest::Latest { to_ts, limit } => {
```
Comment (spb) [inline sable_history/src/pg_history_service.rs:123]: Is there no commonality in these arms that can be broken out? They're quite hard to follow as they are.

File: sable_history/src/pg_history_service.rs [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+                    .collect::<Vec<_>>()
+                }
+            }
+        })
+    }
+}
+
+fn make_historical_event(
+    channel: &crate::models::Channel,
+    (id, timestamp, message_type, text, source_nick, source_ident, source_vhost, source_account): (
```
Comment (spb) [inline sable_history/src/pg_history_service.rs:295]: This might be more readable as a new Queryable struct (possibly with a helper method in the model to query it) - this ties into the comment further up about a helper query method

File: sable_history/src/server/update_handler.rs [hunk 15]
```diff
@@ -131,7 +135,11 @@ impl HistoryServer {
         }
     }
 
-    async fn handle_new_message(&self, new_message: update::NewMessage) -> anyhow::Result<()> {
+    async fn handle_new_message(
+        &self,
+        new_message: update::NewMessage,
+        update_timestamp: i64,
```
Comment (spb) [inline sable_history/src/server/update_handler.rs:141]: Why is this not using the ts field from the Message?

File: sable_network/src/network/event/details.rs [hunk 16]
```diff
@@ -190,10 +190,14 @@ EventDetails => {
     }
 
     #[target_type(ServerId)]
-    struct IntroduceServices {
+    struct IntroduceServicesServer {
         pub sasl_mechanisms: Vec<String>,
     }
 
+    #[target_type(ServerId)]
+    struct IntroduceHistoryServer {
```
Comment (spb) [inline sable_network/src/network/event/details.rs:198]: I realise this is matching the services behaviour and it's probably fine for now, but long term the desired behaviour isn't the same - we need a single authoritative services node, but for history we can have multiple active nodes and just send requests to any of them.

File: sable_history/src/pg_history_service.rs (hunk 13 above)
Comment (progval) [reply sable_history/src/pg_history_service.rs:123]: I can rewrite everything as a Between, I guess

File: sable_history/src/pg_history_service.rs (hunk 14 above)
Comment (progval) [reply sable_history/src/pg_history_service.rs:295]: I don't see what you have in mind

File: sable_history/src/server/update_handler.rs (hunk 15 above)
Comment (progval) [reply sable_history/src/server/update_handler.rs:141]: To make the `@time` tag consistent between echo-message and chathistory https://github.com/Libera-Chat/sable/pull/143/commits/d56db83b2b385dc211a590685b3e1263016f9a91

File: sable_network/src/network/event/details.rs (hunk 16 above)
Comment (progval) [reply sable_network/src/network/event/details.rs:198]: ack

File: sable_history/src/pg_history_service.rs (hunk 13 above)
Comment (progval) [reply sable_history/src/pg_history_service.rs:123]: I managed to deduplicate most of the code: https://github.com/Libera-Chat/sable/pull/143/commits/a8f711b28921d0046392573fcd37af0ca139a4c4
